source/base/function_spherical.cc

```diff
@@ -0,0 +1,303 @@
+      {
+        Assert (sp[0] != 0.,
+                ExcDivideByZero());
+        res += unit_phi * sg[2] / sp[0];
+      }
+
+    return res;
+  }
+
+
+  template <int dim>
+  SymmetricTensor<2,dim>
+  Spherical<dim>::hessian (const Point<dim> &p_,
+                           const unsigned int component) const
+  {
+    SymmetricTensor<2,dim> res;
+
+    const Point<dim> p = p_ - coordinate_system_offset;
+    const std_cxx11::array<double, dim> sp = GeometricUtilities::Coordinates::to_spherical(p);
+    const Tensor<1,dim> sg = sgradient(sp);
+    const SymmetricTensor<2,dim> sh = shessian(sp);
+
+    // FIXME: somewhat backwards, but we need cos/sin's for unit vectors
+    const double cos_theta = std::cos(sp[1]);
+    const double sin_theta = std::sin(sp[1]);
+    const double cos_phi   = std::cos(sp[2]);
+    const double sin_phi   = std::sin(sp[2]);
+    const double r         = sp[0];
+
+    Tensor<1,dim> unit_r, unit_theta, unit_phi;
+    set_unit_vectors(unit_r, unit_theta, unit_phi,
+                     cos_theta, sin_theta,
+                     cos_phi, sin_phi);
+
+    const double sin_phi2 = sin_phi*sin_phi;
+    const double r2 = r*r;
+
+    const double c_utheta2     = r*sg[0]*sin_phi2+cos_phi*sg[2]*sin_phi+sh[1][1];
+    const double c_utheta_ur   = r*sh[0][1]*sin_phi-sg[1]*sin_phi;
+    const double c_utheta_uphi = sh[1][2]*sin_phi-cos_phi*sg[1];
+
+    // const double c_ur_utheta   = (r*sh[0][1]-sg[1])                 * sin_phi;
+    const double c_ur2         = r2*sh[0][0]*sin_phi                * sin_phi;
+    const double c_ur_uphi     = (r*sh[0][2]*sin_phi-sg[2]*sin_phi) * sin_phi;
+
+    // const double c_uphi_utheta = sh[1][2]*sin_phi-cos_phi*sg[1];
+    // const double c_uphi_ur     = r*sh[0][2]*sin_phi2-sg[2]*sin_phi2;
+    const double c_uphi2       = sh[2][2]*sin_phi2+r*sg[0]*sin_phi2;
+
+    Assert (r2*sin_phi2 != 0.,
+            ExcDivideByZero());
+    const double c = 1./(r2*sin_phi2);
+
+    // go through each tensor product
+    SymmetricTensor<2,dim> tmp;
+
+    if (c_utheta2 != 0.)
+      {
+        symm_outer_product(tmp, unit_theta);
+        res += c*c_utheta2 * tmp;
+      }
+
+    if (c_utheta_ur != 0.)
+      {
+        symm_outer_product(tmp, unit_theta, unit_r);
+        res += c*c_utheta_ur * tmp;
+      }
+
+    if (c_utheta_uphi != 0.)
+      {
+        symm_outer_product(tmp, unit_theta, unit_phi);
+        res += c*c_utheta_uphi * tmp;
+      }
+
+    if (c_ur2 != 0.)
+      {
+        symm_outer_product(tmp, unit_r);
+        res += c*c_ur2 * tmp;
+      }
+
+    if (c_ur_uphi != 0.)
+      {
+        symm_outer_product(tmp, unit_r, unit_phi);
+        res += c*c_ur_uphi * tmp;
+      }
+
+    if (c_uphi2 != 0.)
+      {
+        symm_outer_product(tmp, unit_phi);
+        res += c*c_uphi2 * tmp;
+      }
+
+    return res;
+  }
+
+
+  template <int dim>
+  std::size_t
+  Spherical<dim>::memory_consumption () const
+  {
+    return coordinate_system_offset.memory_consumption();
```

Comment:
That's not quite correct because it neglects the memory consumption of the base class. I think the correct return value is `sizeof(Spherical<dim>)`.